src/TransformationUtils.cc

```diff
@@ -88,57 +88,63 @@ std::unique_ptr<TransitionSystem> toTransitionSystem(ChcDirectedGraph const & gr
     return ts;
 }
 
-bool strongConnection(std::unordered_set<int> & visitedVertices, std::unordered_set<int> & verticesOnStack,
+// This function follows a Tarjan's strong connection detection algorithm:
+// https://en.wikipedia.org/wiki/Tarjan's_strongly_connected_components_algorithm
+// If some vertice is added on stack and it was already visited, then the strong connection is detected
+// It was updated to remember the nodes which participate in the loop (lines 123, 127)
+void strongConnection(std::unordered_set<SymRef, SymRefHash> & visitedVertices,
```

Comment:
Since you do not implement full Tarjan algorithm, but only detect one loop, maybe the function should have a different name?

I am thinking that you could create a `FindFirstLoop` class and move the implementation there.
Then the method could be just `visit` for visiting a node.

Reply:
Updated